---
title: "BLOSUM and Trees"
output:
  html_document: default
  word_document: default
  pdf_document: default
---

## Configure environment

For more information on the BiocManager::install function, see https://cran.r-project.org/web/packages/BiocManager/BiocManager.pdf

You can run each code segment independently by clicking on the small
green triangle in the upper right of each code section. Try it below as
you set up your environment (with supporting software packages).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require("ape", quietly = TRUE))
    install.packages("ape")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (!require("phangorn", quietly = TRUE))
    install.packages("phangorn")
if (!require("adegenet", quietly = TRUE))
    install.packages("adegenet")
if (!require("Biostrings", quietly = TRUE))
    suppressWarnings(BiocManager::install(c("Biostrings","ggtree"),update=FALSE))
library(Biostrings)
library(ggtree)
library(ape)
library(phangorn)
library(adegenet)
data(BLOSUM62)
```

## Exercise 1. Understanding the BLOSUM scoring matrix.

Look at https://en.wikipedia.org/wiki/BLOSUM

In the following code segment, modify the subject_example sequence by creating similar subject_better and subject_worse sequences that will increase versus decrease the resulting BLOSUM62-based alignment score.

(Note this exercise is adapted from https://bioconductor.org/packages/devel/bioc/manuals/Biostrings/man/Biostrings.pdf)

```{r blosum_example, echo=FALSE}
pattern <- c("TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQAT",
"GGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAG",
"SRPNRFAPTLMSSCITSTTGPPAWAGDRSHE")
subject_example <- c("TSPASIRPPAGPSSRRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGT",
"CTTSTSTRHRGRSGWRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTT",
"GPPAWAGDRSHE")
subject_worse <- c("WRASRKSPPAGPSSRRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGT",
"CTTSTSTRHRGRSGWRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTT",
"GPPAWAGDRSHETSPASIRG")
subject_better <- c("TSPASIRPPAGPSSRRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGT",
"GGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAG",
"SRPNRFAPTLMSSCITSTTGPPAWAGDRSHE")

pattern <- unlist(AAStringSet(pattern))
subject_example <- unlist(AAStringSet(subject_example))
subject_worse <- unlist(AAStringSet(subject_worse))
subject_better <- unlist(AAStringSet(subject_better))
pattern # original pattern
subject_example # original subject
data(BLOSUM62)
pa5 <- pairwiseAlignment(pattern, subject_example,
substitutionMatrix=BLOSUM62,
gapOpening=9.5, gapExtension=0.5)
print(pa5)
pa5 <- pairwiseAlignment(pattern, subject_better,
substitutionMatrix=BLOSUM62,
gapOpening=9.5, gapExtension=0.5)
print(pa5)
pa5 <- pairwiseAlignment(pattern, subject_worse,
substitutionMatrix=BLOSUM62,
gapOpening=9.5, gapExtension=0.5)
print(pa5)
```

## Levenshtein distance example

Look at the manual for stringDist (https://rdrr.io/bioc/Biostrings/man/stringDist.html) and write a function that calculate the Levenshtein scores for the pattern sequence versus subject_example versus subject_better versus subject_worse.

```{r levenshtein_example, echo=FALSE}
# You'll need to add subject_better and subject_worse to the vector
# in the c() part of the function below.
distance_matrix <- 
  stringDist(c(as.character(pattern),
             as.character(subject_example),
             as.character(subject_better),
             as.character(subject_worse)
             ),
           method="levenshtein")
distance_matrix
```

Wow, that is a distance matrix! This can be used to create a tree. Save the value of the above function to a variable, and then put that variable into the hclust function.
(Remember though that cluster-based methods do not produce trees that are 
accurate.)

```{r cluster_based_tree}
plot(
  hclust(distance_matrix,method="average"),
  labels=c("pattern","example","better","worse")
)

```

## Phylogenetic Trees

# Simulated phylogeny

```{r simulated_tree}
plot(rtree(n=20))
```

# The T7 sequence data are from Hillis et al seminal work in the 1990s.

```{r phylogenetic_tree}
t7 <- readDNAStringSet("t7hillis9seq_fasta2.txt")

protein03 <- subseq(t7,1,519)
protein17 <- subseq(t7,520,875)
protein18 <- subseq(t7,876,1091)
```

Think about how the Levenshtein scores work. Does a high Levenshtein
score mean very similar or very different?

You can analyze either the t7 supergene sequence (1091 nucleotides)
or some of the individual protein sequences. Which phage lineages seem most
closely related from the analysis?

```{r distancematrix_t7}
distance_matrix_t7 <-
  stringDist(t7, method="levenshtein")
distance_matrix_t7
```

```{r cluster_based_tree_t7}
plot(
  hclust(distance_matrix_t7,method="average"),main="UPGMA tree"
)
```

Adapted from
https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html

Most phylogenetic trees require a distance matrix as input data from
which the tree reconstruction algorithm works.

```{r distance_matrix_t7}
D <- dist.dna(as.DNAbin(t7), model = "TN93")
temp <- as.data.frame(as.matrix(D))
table.paint(temp, cleg=0, clabel.row=.5, clabel.col=.5)
```

Here is an example of building a neighbor joining tree.

```{r neighbor_joining_t7}
tre1 <- nj(D)
plot(tre1)
str(tre1)
tre1$tip.label
plot(root(tre1,out=9))
```

## Here is an example of building a parsimony tree.

```{r parsimony_t7}
D_raw <- dist.dna(as.DNAbin(t7), model = "raw")
tre2 <- nj(D_raw)
tre2.pars <- optim.parsimony(tre2, as.phyDat(as.DNAbin(t7)))

tre2.pars$tip.label # figure out which tip label is R which should be the outgroup
plot(root(tre2.pars,out=7))
```

## Here is an example of building a maximum likelihood tree.

```{r maximumlikelihood_t7}
fit.ini1 <- pml(tre1, as.phyDat(as.DNAbin(t7)), k=4)

treML <- optim.pml(fit.ini1, optNni=TRUE, optBf=TRUE, optQ=TRUE, optGamma=TRUE)

treML$tree$tip.label # figure out which tip label is R which should be the outgroup

plot(root(treML$tree,out=9))
```

## Final Exercise

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document.

Try knitting both an HTML document and an MS word document. Submit your MS word document through blackboard for today's assignment.

